File: Semester4/Systems for Design and Implementation/mpp-proiect-csharp-raulfrandes/persistence/dbRepositories/TrialDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using log4net;
using model;
using model.validator;
using persistence.interfaces;

namespace persistence.dbRepositories;

public class TrialDbRepository: ITrialRepository
{
    private static readonly ILog Log = LogManager.GetLogger("TrialDBRepository");

    private IDictionary<string, string> _props;
    private IValidator<Trial> _trialValidator;

    public TrialDbRepository(IDictionary<string, string> props, IValidator<Trial> trialValidator)
    {
        Log.Info("Creating TrialDBRepository");
        this._props = props;
        this._trialValidator = trialValidator;
    }

    private Referee? FindReferee(long idReferee)
    {
        Log.InfoFormat("Entering FindReferee with value {0}", idReferee);
        IDbConnection connection = DbUtils.GetConnection(_props);

        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from referees where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = idReferee;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        String name = dataReader.GetString(1);
                        String username = dataReader.GetString(2);
                        String password = dataReader.GetString(3);
                        Referee referee = new Referee(name, username, password);
                        referee.Id = idReferee;
                        Log.InfoFormat("Exiting FindReferee with value {0}", referee);
                        return referee;
                    }
                    Log.InfoFormat("Exiting FindReferee with value {0}", null);
                    return null;
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }
    
    public IEnumerable<Trial> FindAll()
    {
        Log.InfoFormat("Entering FindAll");
        IDbConnection connection = DbUtils.GetConnection(_props);

        IList<Trial> trials = new List<Trial>();
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from trials";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        string description = dataReader.GetString(2);
                        long idReferee = dataReader.GetInt64(3);
                        Referee referee = FindReferee(idReferee)!;
                        Trial trial = new Trial(name, description, referee);
                        trial.Id = id;
                        trials.Add(trial);
                    }
                }
            }

            Log.InfoFormat("Exiting FindAll with value {0}", trials);
            return trials;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Trial? FindById(long entityId)
    {
        Log.InfoFormat("Entering FindById with value {0}", entityId);
        IDbConnection connection = DbUtils.GetConnection(_props);

        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select * from trials where id = @id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entityId;
                command.Parameters.Add(paramId);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        long id = dataReader.GetInt64(0);
                        string name = dataReader.GetString(1);
                        string description = dataReader.GetString(2);
                        long idReferee = dataReader.GetInt64(3);
                        Referee referee = FindReferee(idReferee)!;
                        Trial trial = new Trial(name, description, referee);
                        trial.Id = id;
                        Log.InfoFormat("Exiting FindById with value {0}", trial);
                        return trial;
                    }
                    Log.InfoFormat("Exiting FindById with value {0}", null);
                    return null;
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Trial? Save(Trial entity)
    {
        _trialValidator.Validate(entity);
        Log.InfoFormat("Entering Save with value {0}", entity);
        IDbConnection connection = DbUtils.GetConnection(_props);
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "insert into trials (name, description, id_referee) values (@name, @description, @id_referee)";

                IDbDataParameter paramName = command.CreateParameter();
                paramName.ParameterName = "@name";
                paramName.Value = entity.Name;
                command.Parameters.Add(paramName);

                IDbDataParameter paramDescription = command.CreateParameter();
                paramDescription.ParameterName = "@description";
                paramDescription.Value = entity.Description;
                command.Parameters.Add(paramDescription);

                IDbDataParameter paramReferee = command.CreateParameter();
                paramReferee.ParameterName = "id_referee";
                paramReferee.Value = entity.Referee.Id;
                command.Parameters.Add(paramReferee);

                var rows = command.ExecuteNonQuery();
                Log.InfoFormat("saved {0} instances", rows);

                if (rows > 0)
                {
                    command.CommandText = "select last_insert_rowid()";
                    var result = command.ExecuteScalar();
                    if (result != null && result != DBNull.Value)
                    {
                        long id = Convert.ToInt64(result);
                        entity.Id = id;
                        Log.InfoFormat("Exiting Save with value {0}", entity);
                        return entity;
                    }
                }
            }
            Log.InfoFormat("Exiting Save with value {0}", null);
            return null;
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    public Trial DeleteById(long entityId)
    {
        throw new NotImplementedException();
    }

    public Trial Update(Trial entity)
    {
        throw new NotImplementedException();
    }
}

File: Semester4/Systems for Design and Implementation/mpp-proiect-java-raulfrandes/TriathlonPersistence/src/main/java/triathlon/persistence/dbRepositories/TrialDBRepository.java
package triathlon.persistence.dbRepositories;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import triathlon.model.Referee;
import triathlon.model.Trial;
import triathlon.model.validator.Validator;
import triathlon.persistence.interfaces.ITrialRepository;
import triathlon.persistence.RepositoryException;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class TrialDBRepository implements ITrialRepository {
    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();
    private final Validator<Trial> trialValidator;

    public TrialDBRepository(Properties props, Validator<Trial> trialValidator) {
        logger.info("Initializing TrialDBRepository with properties: {}", props);
        this.dbUtils = new JdbcUtils(props);
        this.trialValidator = trialValidator;
    }

    private Referee findReferee(Long idReferee) {
        if (idReferee == null) {
            throw new RepositoryException("Referee cannot be null");
        }
        logger.traceEntry("searching referee with id {}", idReferee);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from referees where id = ?")) {
            preparedStatement.setLong(1, idReferee);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                String name = resultSet.getString("name");
                String username = resultSet.getString("username");
                String password = resultSet.getString("password");
                Referee referee = new Referee(name, username, password);
                referee.setId(idReferee);
                logger.traceExit(referee);
                return referee;
            }
            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Iterable<Trial> findAll() {
        logger.traceEntry();
        Connection connection = dbUtils.getConnection();
        List<Trial> trials = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection
                .prepareStatement("select * from trials")){
            ResultSet resultSet = preparedStatement.executeQuery();
            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String description = resultSet.getString("description");
                Long idReferee = resultSet.getLong("id_referee");
                Referee referee = findReferee(idReferee);
                Trial trial = new Trial(name, description, referee);
                trial.setId(id);
                trials.add(trial);
            }
            logger.traceExit(trials);
            return trials;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Trial findById(Long entityId) {
        if (entityId == null) {
            throw new RepositoryException("Id cannot be null");
        }
        logger.traceEntry("searching trial with id {}", entityId);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection
                .prepareStatement("select * from trials where id = ?")) {
            preparedStatement.setLong(1, entityId);
            ResultSet resultSet = preparedStatement.executeQuery();
            if (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String name = resultSet.getString("name");
                String description = resultSet.getString("description");
                Long idReferee = resultSet.getLong("id_referee");
                Referee referee = findReferee(idReferee);
                Trial trial = new Trial(name, description, referee);
                trial.setId(id);
                logger.traceExit(referee);
                return trial;
            }
            logger.traceExit(null);
            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Trial save(Trial entity) {
        if (entity == null) {
            throw new RepositoryException("Entity cannot be null");
        }
        trialValidator.validate(entity);
        logger.traceEntry("saving trial {}", entity);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection
                .prepareStatement("insert into trials (name, description, id_referee) values (?, ?, ?)",
                        Statement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setString(1, entity.getName());
            preparedStatement.setString(2, entity.getDescription());
            preparedStatement.setLong(3, entity.getReferee().getId());
            int result = preparedStatement.executeUpdate();
            logger.trace("saved {} instances", result);

            ResultSet resultSet = preparedStatement.getGeneratedKeys();
            if (resultSet.next()) {
                Long id = resultSet.getLong(1);
                entity.setId(id);
                logger.traceExit(entity);
                return entity;
            }

            return null;
        } catch (SQLException e) {
            logger.error(e);
            throw new RepositoryException("Error DB " + e);
        }
    }

    @Override
    public Trial deleteById(Long entityId) {
        return null;
    }

    @Override
    public Trial update(Trial entity) {
        return null;
    }
}
